### ISSUE TITLE
[Feature] Allow multiple terminals opened in different tabpages

### ISSUE THREAD

**Issue by Bekaboo** [akinsho/toggleterm.nvim#452]
Hi, thanks for making this awesome plugin, it makes nvim's builtin terminal much more ergonomic and user-friendly.

Sometimes I work on 2+ projects in neovim and keeps them in multiple tabs, and it would be nice to have multiple toggleterms opened in these tabs. Currently if I open a toggleterm in one tab, switch to another tab and call 'toggleterm.toggle_command()` again, toggleterm will close the terminal in the previous tabpage instead of opening a new one in the current tabpage.

I assume this might be a config issue and can be solved by passing proper opts to toggleterm? The following is my config if that's related:

```lua
local toggleterm = require('toggleterm')

toggleterm.setup({
  open_mapping = '<C-\\><C-\\>',
  shade_terminals = false,
  start_in_insert = false,
  hide_numbers = true,
  autochdir = false,
  persist_size = false,
  direction = 'horizontal',
  float_opts = {
    border = 'shadow',
    width = function()
      return math.floor(vim.go.columns * 0.7)
    end,
    height = function()
      return math.floor(vim.go.lines * 0.7)
    end,
    winblend = 0,
  },
  size = function(term)
    if term.direction == 'horizontal' then
      return vim.go.lines * 0.3
    elseif term.direction == 'vertical' then
      return vim.go.columns * 0.35
    end
  end,
  highlights = {
    NormalFloat = { link = 'NormalFloat' },
    FloatBorder = { link = 'FloatBorder' },
    WinBarActive = { link = 'WinBar' },
    WinBarInactive = { link = 'WinBarNC' },
    MatchParen = { link = 'None' },
  },
  winbar = {
    enabled = false,
  },
  on_open = function()
    vim.cmd('silent! normal! 0')
  end,
})
```

**Comment by akinsho** [comment]
Hi @Bekaboo,

Tab scoped terminals is definitely something that has come up before and is an interesting idea. There are definitely some things to iron out about how it would work though. The simplest solution is that if you set some value on the terminal itself and if true then toggleterm treats it as only existing in the tab it was created in and if false it opens wherever.

I worry about introducing all of this complexity though as one of the first things I imagine this will introduce is a bunch of feature requests to accomodate a bunch of minutiae. I tend to lean more towards keeping the plugin as simple as possible to reduce the maintenance burden in the long run.

TLDR: seems like a cool idea, not clear on all the specifics of how it would work. Don't currently have the time to build it and I'm worried it will lead to more complexity.

**Comment by Bekaboo** [comment]
OK, thanks for your response. should I close this issue?

**Comment by akinsho** [comment]
@Bekaboo no we can leave it open, someone might decide pitch in since it's a popular idea I've heard before. If not maybe when things calm down at work I'll take a look at it.

**Comment by coot** [comment]
It would be something really handy.  I often use multiple tabs for different projects, and `toggleterm` is not very intuitive when closing a terminal in another tab changing focus to another tab page.

**Comment by akinsho** [comment]
It's fine to want things but if no one *else* is willing to put the time or effort in or contribute since this is OSS it won't get done. For better or worse I have a life and a job and this isn't a priority for me over those. Please no more plus one comments.

**Comment by anurag3301** [comment]
I was also recommending an easy way to split the terminal and have more. Like in below example I tried splitting it with `<C-W>v`
![image](https://github.com/akinsho/toggleterm.nvim/assets/52702259/70aeb116-0842-4e15-b32c-f22e958d9350)
It does seem like we have tow terminals, but they are just copied, If I write to one, it appears on the second to. Feels like could be a nice features.
![image](https://github.com/akinsho/toggleterm.nvim/assets/52702259/f83e4cee-1ebf-404f-a510-c2df59b646fb)

**Comment by tarunaksha** [comment]
> I was also recommending an easy way to split the terminal and have more. Like in below example I tried splitting it with `<C-W>v` ![image](https://private-user-images.githubusercontent.com/52702259/296793470-70aeb116-0842-4e15-b32c-f22e958d9350.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY0NDY2MzQsIm5iZiI6MTcwNjQ0NjMzNCwicGF0aCI6Ii81MjcwMjI1OS8yOTY3OTM0NzAtNzBhZWIxMTYtMDg0Mi00ZTE1LWIzMmMtZjIyZTk1OGQ5MzUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI4VDEyNTIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkMjYzNGIzNTVhYzYwOWNlNDA3OWUxY2I1NDg1ZjhiODk2ZTgyZjc0MjkxZTFmOTliNTk0NjE5YTE1YmYwNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.gK-eR65KQgGJ1TtytlLkdZZ38zdJfJd5u6RxVaCixJQ) It does seem like we have tow terminals, but they are just copied, If I write to one, it appears on the second to. Feels like could be a nice features. ![image](https://private-user-images.githubusercontent.com/52702259/296793784-f83e4cee-1ebf-404f-a510-c2df59b646fb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY0NDY2MzQsIm5iZiI6MTcwNjQ0NjMzNCwicGF0aCI6Ii81MjcwMjI1OS8yOTY3OTM3ODQtZjgzZTRjZWUtMWViZi00MDRmLWE1MTAtYzJkZjU5YjY0NmZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTI4VDEyNTIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk0M2YzMDgzOTViYTNhMTkwZTA0MWFhMmQwMGM1NmFhN2Q5MmM0MDY2OWJhMWQxMjA0ODUwY2NhZDY3N2RiZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HX-Gp89xxNYUmXrOIZhzWzynJn8oNaPr6JVTefWsp0w)

I have  a solution here....you can check....In my case each terminal is working fine >> https://github.com/akinsho/toggleterm.nvim/issues/533#issuecomment-1913585725

**Comment by erlangparasu** [comment]
https://github.com/akinsho/toggleterm.nvim/issues/593

**Comment by radgeRayden** [comment]
Here's an implementation of this feature that I wrote for my init.lua. For my usecase I want to have a per tab "bottom terminal" that persists its state. I hadn't seen the note about `find_open_windows` above so maybe it could be simplified a bit by reusing some of toggleterm's functions.

```lua
local bottom_terminals = {}
local function open_bottom_terminal()
  local Terminal = require('toggleterm.terminal').Terminal
  local tabpage = vim.api.nvim_get_current_tabpage()
  local tab_term = bottom_terminals[tabpage]

  if not tab_term then
    bottom_terminals[tabpage] = Terminal:new {
      go_back = false,
    }
    tab_term = bottom_terminals[tabpage]
  end

  -- we need to close all other terminals, otherwise this instance will try to join any other instance that is also in horizontal mode
  for _,term in ipairs(bottom_terminals) do
    if term ~= tab_term and term:is_open() then
      term:toggle()
    end
  end

  -- toggling a term sets tab focus, so we need to steal it back
  vim.api.nvim_set_current_tabpage(tabpage)
  tab_term:toggle()
end
```

I set `<Leader>tt` to run this function for toggling the terminal, and have a `<C-w>` mapping to hide the terminal without closing it.

**Comment by dreamvibe1993** [comment]
> I was also recommending an easy way to split the terminal and have more. Like in below example I tried splitting it with `<C-W>v` ![image](https://github.com/akinsho/toggleterm.nvim/assets/52702259/70aeb116-0842-4e15-b32c-f22e958d9350) It does seem like we have tow terminals, but they are just copied, If I write to one, it appears on the second to. Feels like could be a nice features. ![image](https://github.com/akinsho/toggleterm.nvim/assets/52702259/f83e4cee-1ebf-404f-a510-c2df59b64

I also experience the same thing. Anybody else? Any advises/workarounds?

**Comment by Bsd-us** [comment]
+1 for this feature I hope this gets accepted, it feels like a requirement to use tabs !

**Comment by axelhj** [comment]
If you open a horizontal toggle-terminal and then use `:ToggleTerm2`, a new terminal will open as a split terminal "in tow" besides the existing horizontal terminal. I use mappings for that.

It would take some extra scripting to do that from a single keymap though, which I haven't tried. @dreamvibe1993 Might help to do what you want to do.

**Comment by apple1417** [comment]
I've been using the following auto commands to get a basic form of this:
```lua
vim.api.nvim_create_autocmd("TabEnter", {
    group = vim.api.nvim_create_augroup("~toggleterm-tab-enter", { clear = true }),
    callback = function()
        local ok, info = pcall(vim.api.nvim_tabpage_get_var, 0, "~toggleterm")
        if not ok or not info then
            info = {
                was_open = false,
                view = { terminals = {}, focus_term_id = nil },
            }
        end

        local ui = require("toggleterm.ui")
        ui.save_terminal_view(info.view.terminals, info.view.focus_term_id)
        if info.was_open then
            ui.open_terminal_view()
        end
    end,
})

vim.api.nvim_create_autocmd("TabLeave", {
    group = vim.api.nvim_create_augroup("~toggleterm-tab-leave", { clear = true }),
    callback = function()
        local ui = require("toggleterm.ui")

        local any_open, windows = ui.find_open_windows()
        if any_open and windows then
            ui.close_and_save_terminal_view(windows)
        end

        vim.api.nvim_tabpage_set_var(0, "~toggleterm", {
            was_open = any_open,
            view = ui.get_terminal_view(),
        })
    end,
})
```
This saves the layout per tab, you can still open the same terminal in multiple tabs, same as how windows work. Notably, if you've never opened a terminal in a given tab, the bind will default to re-opening terminal 1. If you want a new terminal, you'll instead need to remember "tab 1 had terminals 1+3, tab 2 had terminal 2, so in tab 3 press `4<C-\>`". Once your terminals are set up I've found it pretty seamless though.
